### PR TITLE
Dropdown: refine handling of keydown in input and textarea

### LINKED ISSUE #21366: Dropdown: don't ignore keydown from input or textarea
I'm reopening an issue for v4, as #15757 got [closed without resolution for v3](https://github.com/twbs/bootstrap/issues/15757#event-778679826).

At the time I opened the issue, [v4 was still filtering key down events for input or textarea](https://github.com/twbs/bootstrap/blob/3c12d73203bad27edcaa1072e263fdab2a2c14f7/js/src/dropdown.js#L214-L218):
```
static _dataApiKeydownHandler(event) {
  if (!/(38|40|27|32)/.test(event.which) ||
     /input|textarea/i.test(event.target.tagName)) {
    return
  }
```

In my opinion, a common use of an `<input type="text">` and dropdown combination would be displaying search results as a dropdown ala http://developer.github.com:
image

In that respect, can you please make the behavior configurable (if you don't want to revert the default behavior).

Thank you.

### PR DESCRIPTION
**Pull request by pvdlg**
Fixes #21366

This PR handle better the key inputs in `input` and `textarea` within `dropdown` and `dropdown-menu`.
As a side note, keyboard navigation is current broken and the fix is here #21535. 

**This PR will be fully working only when #21535 lands.**

Here is the new behaviors.

#### `input` and `textarea` within a `dropdown-menu`:
- **Escape key:** dropdown behavior => close the menu
- **Arrow down:** default behavior (move to the cursor to the end of the text in the `input` or `textarea`)
- **Arrow up:** default behavior => move to the cursor to the beginning of the text in the `input` or `textarea`
- **Space key:** default behavior => enter a space character in the `input` or `textarea`
- **Other key:** default behavior => enter the character in the `input` or `textarea`

#### `input` and `textarea` within a `dropdown` but outside of the `dropdown-menu`:
- **Escape key:** dropdown behavior => close the menu
- **Arrow down:** dropdown behavior => put the focus on the first/next `dropdown-item` in the menu
- **Arrow up:** dropdown behavior => put the focus on the first/previous `dropdown-item` in the menu
- **Space key:** default behavior => enter a space character in the `input` or `textarea`
- **Other key:** default behavior => enter the character in the `input` or `textarea`

### CODE REVIEW

**Review by Johann-S — APPROVED** [review]
LGTM 👍 

But we have to review this #21535 first

**Comment by Johann-S** [comment]
Can you update your branch @vanduynslagerp ? Thank you 👍

**Comment by Johann-S** [comment]
Can you fix the conflict @vanduynslagerp ? please 👍

**Comment by Johann-S** [comment]
Your unit test failed because of the lack of `.dropdown-item` and `li` isn't focusable without `tabindex="1"`

With this HTML everything is fine : 

```html
      <ul class="tabs">
        <li class="dropdown">
          <input type="text" id="input" data-toggle="dropdown">
          <ul class="dropdown-menu" role="menu">
            <li id="item1" tabindex="1" class="dropdown-item"><a href="#">Secondary link</a></li>
            <li id="item2" tabindex="1" class="dropdown-item"><a href="#">Something else here</a></li>
            <li class="divider"/>
            <li tabindex="1" class="dropdown-item"><a href="#">Another link</a></li>
          </ul>
        </li>
      </ul>

**Review by Johann-S — CHANGES_REQUESTED** [review]
Your unit test failed because of the lack of `.dropdown-item` and `li` isn't focusable without `tabindex="1"`

With this HTML everything is fine : 

```html
      <ul class="tabs">
        <li class="dropdown">
          <input type="text" id="input" data-toggle="dropdown">
          <ul class="dropdown-menu" role="menu">
            <li id="item1" tabindex="1" class="dropdown-item"><a href="#">Secondary link</a></li>
            <li id="item2" tabindex="1" class="dropdown-item"><a href="#">Something else here</a></li>
            <li class="divider"/>
            <li tabindex="1" class="dropdown-item"><a href="#">Another link</a></li>
          </ul>
        </li>
      </ul>

**Comment by mdo** [comment]
@Johann-S I pushed the class name change but tests are still failing. Any chance you could take a look?

**Comment by Johann-S** [comment]
Superseded by : #24557

File: js/src/dropdown.js [hunk 1]
```diff
@@ -357,8 +357,17 @@ const Dropdown = (() => {
     }
 
     static _dataApiKeydownHandler(event) {
-      if (!REGEXP_KEYDOWN.test(event.which) || /button/i.test(event.target.tagName) && event.which === SPACE_KEYCODE ||
-         /input|textarea/i.test(event.target.tagName)) {
+      // If not input/textarea:
+      //  - And not a key in REGEXP_KEYDOWN => not a dropdown command
+      // If input/textarea:
+      //  - If space key => not a dropdown command
+      //  - If key is other than excape
```
Comment (XhmikosR) [inline js/src/dropdown.js]: `escape`

**Pull request closed without merging**